implement WebSocket API

Witcher01 opened the pull request:
following the esp_websocket_client found here[^1][^2], this commit
wraps the mentioned C library in Rust to expose a safe WebSocket
interface.
The exposed API copies the previously implemented MQTT API. On creating
a `EspWebSocketClientConfig` and passing it to a `EspWebSocketClient`,
the user can call `send` to send data. Manual sending of fragmented data,
Ping/Pong, or Connect/Disconnect are not supported due to the underlying
C library not supporting this, either.
Receiving is done in a new thread with the returned
`EspWebSocketConnection` and its blocking `next` function call, which
wakes on receiving one of the events defined in `WebSocketEventType`.
The `EspWebSocketClient` takes a `time::Duration` timeout since the
underlying C lib requires this but the defined interface in the
`embedded_svc` create does not provide it as a parameter.
On dropping the `EspWebSocketClient` the C functions
`esp_websocket_client_close` and `esp_websocket_client_destroy` are
being called, cleanly closing the connection to the WebSocket server.

[^1]: https://espressif.github.io/esp-protocols/esp_websocket_client/
[^2]: https://github.com/espressif/esp-protocols/

Witcher01 committed:
implement WebSocket API

following the esp_websocket_client found here[^1][^2], this commit
wraps the mentioned C library in Rust to expose a safe WebSocket
interface.
The exposed API copies the previously implemented MQTT API. On creating
a `EspWebSocketClientConfig` and passing it to a `EspWebSocketClient`,
the user can call `send` to send data. Manual sending of fragmented data,
Ping/Pong, or Connect/Disconnect are not supported due to the underlying
C library not supporting this, either.
Receiving is done in a new thread with the returned
`EspWebSocketConnection` and its blocking `next` function call, which
wakes on receiving one of the events defined in `WebSocketEventType`.
The `EspWebSocketClient` takes a `time::Duration` timeout since the
underlying C lib requires this but the defined interface in the
`embedded_svc` create does not provide it as a parameter.
On dropping the `EspWebSocketClient` the C functions
`esp_websocket_client_close` and `esp_websocket_client_destroy` are
being called, cleanly closing the connection to the WebSocket server.

[^1]: https://espressif.github.io/esp-protocols/esp_websocket_client/
[^2]: https://github.com/espressif/esp-protocols/

elwerene commented on src/lib.rs:
```diff
@@ -83,4 +83,7 @@ pub mod timer;
 ))]
 pub mod wifi;
 
+// TODO: cfg
```
das braucht alloc, dann vielleicht einfach dieses feature hier aufnehmen

elwerene commented on src/ws/client.rs:
```diff
@@ -0,0 +1,533 @@
+    fn drop(&mut self) {
+        let mut message = self.state.message.lock();
+
+        if message.is_some() {
```
ich würde `message.take().is_some()` benutzen, das spart eine Zeile Code

Witcher01 replied:
Verstehe nicht genau, wie du meinst 🤔

elwerene replied:
```
    fn drop(&mut self) {
        let mut message = self.state.message.lock();

        if message.take().is_some() {
            self.state.processed.notify_all();
        }
      }
      ```

elwerene replied:
siehe https://doc.rust-lang.org/std/option/enum.Option.html#method.take

Witcher01 replied:
Achso, ich war logisch bei ner anderen Methode. Danke

elwerene commented on src/ws/client.rs:
```diff
@@ -0,0 +1,533 @@
+    pub reconnect_timeout_ms: time::Duration,
+    pub network_timeout_ms: time::Duration,
+    pub ping_interval_sec: time::Duration,
+    // TODO: pub if_name:
```
was ist mit diesen todos? Ist das viel Aufwand, diese zu implementieren?

Witcher01 replied:
Hm, aus irgendeinem Grund hab ich gedacht die werden nicht unterstützt... Die implementiere ich noch

elwerene commented on src/ws/client.rs:
```diff
@@ -0,0 +1,533 @@
+            // default keep_alive_* values are overwritten later in this function
+            ping_interval_sec: conf.ping_interval_sec.as_secs() as _,
+
+            // TODO if_name: *mut ifreq,
```
das todo ist nicht klar genug

elwerene commented on src/ws/client.rs:
```diff
@@ -0,0 +1,533 @@
+    fn post(&self, event_id: i32, event: *mut esp_websocket_event_data_t) {
+        let mut message = self.0.message.lock();
+
+        while message.is_some() {
```
kann das jemals vorkommen, dass message hier some ist?

Witcher01 replied:
Ne hast recht, das Semaphore wird beim Empfangen einer Nachricht genommen und erst freigegeben, wenn der Callback endet bzw. wenn die Nachricht verarbeitet wurde

elwerene commented:
Ich würde noch dazu schreiben, dass dieser merge request den anderen braucht. also sobald du es gegen die esp-rs repos gestellt hast

Witcher01 commented:
Alles klar, mach ich dann

ivmarkov committed:
Set the url in the config so that mqtt ssl validation passes

ivmarkov committed:
Merge branch 'master' of https://github.com/esp-rs/esp-idf-svc

ivmarkov committed:
Untangle executors from signals

ivmarkov committed:
Mqtt client & its connection can now be dropped in any order

Witcher01 committed:
Add missing fields in `EspWebSocketClientConfig`

Missing fields for the used interface and SSL data have been added.
`From` on `(esp_websocket_client_config_t, RawCstrs)` had to be
converted to a `TryFrom` as the used interface field should be a 6 ASCII
character long array, but cannot be validated anywhere else.
Some returned Errors have been converted to `anyhow::Error` types for
simplicity.

ivmarkov committed:
async executor traits

ivmarkov committed:
Formatting fixes

Witcher01 committed:
Fix bug receiving message assumed to end in '\0'

ivmarkov committed:
Minor improvements

ivmarkov committed:
More robust mqtt connection impl

ivmarkov committed:
Separate Unblocking from natural async wrapper

ivmarkov committed:
Unblocker-free impls, using type-alias-impl-trait

ivmarkov committed:
Move the async stuff in a separate module

ivmarkov committed:
Saner names; simplify the message API

ivmarkov committed:
Saner names; simplify the message API

ivmarkov committed:
New release

ivmarkov committed:
Update to latest embuild

ivmarkov committed:
clippy

ivmarkov committed:
Compatibility with ESP-IDF V5

ivmarkov committed:
Compatibility with ESP-IDF V5

ivmarkov committed:
CMake build fixes

gyscos committed:
Fix ambiguous `log` import (#64)

Otherwise it may conflict with the one imported from esp_idf_sys.

```
error[E0659]: `log` is ambiguous
 --> /esp-idf-svc-0.39.1/src/sntp.rs:4:5
  |
4 | use log::*;
  |     ^^^ ambiguous name
  |
  = note: ambiguous because of multiple potential import sources
  = note: `log` could refer to a crate passed with `--extern`
  = help: use `::log` to refer to this crate unambiguously
note: `log` could also refer to the struct imported here
 --> /esp-idf-svc-0.39.1/src/sntp.rs:8:5
  |
8 | use esp_idf_sys::*;
  |     ^^^^^^^^^^^^^^
  = help: use `self::log` to refer to this struct unambiguously

For more information about this error, try `rustc --explain E0659`.
```

callmephilip committed:
add support for buffer_size_tx config option

MabezDev committed:
Merge pull request #72 from bakery/http-client-buffer-size-tx

Add support for buffer_size_tx config option in HTTP Client

Witcher01 committed:
annotate unused field, remove unused import

annotate the unused callback field in `EspWebSocketClient` to allow dead
code. it exists so that the callback is not dropped, as it is used
outside of the scope it is instantiated in (in the c lib).
remove an unused import

Witcher01 committed:
temporarily enable experimental feature for embedded-svc

Witcher01 committed:
adjust for default ESP-IDF version being 4.3.2

ESP-IDF v4.3.2 has a few differences to v4.4, which are addressed here
by conditional compilation.

lamperez committed:
Fix panic with zero-length MQTT messages (#77)

ivmarkov committed:
Don't use dyn Any

Witcher01 committed:
cfg: remove 'all' for websocket

Witcher01 committed:
derive usual traits on EspWebSocketTransport

Witcher01 committed:
use EspError instead of anyhow

Witcher01 committed:
restructure EspWebSocketClientConfig

take a mandatory uri on creating a new EspWebSocketClient. this makes
the fields uri, host, port, and path on the config unnecessary

Witcher01 committed:
task_prio and task_stack are usize

Witcher01 committed:
use only one condvar instead of two

ivmarkov committed:
Implement Send

ivmarkov committed:
Changes to async executor

ivmarkov committed:
Merge branch 'master' into websocket-implementation
